src/db.rs

```diff
@@ -215,30 +215,34 @@ impl DbInner {
 		self.columns[col as usize].get_size(&key, log)
 	}
 
-	// Commit is simply adds the the data to the queue and to the overlay and
+	// Commit simply adds the the data to the queue and to the overlay and
 	// exits as early as possible.
 	fn commit<I, K>(&self, tx: I) -> Result<()>
-		where
-			I: IntoIterator<Item=(ColId, K, Option<Value>)>,
-			K: AsRef<[u8]>,
+	where
+		I: IntoIterator<Item=(ColId, K, Option<Value>)>,
+		K: AsRef<[u8]>,
 	{
-		{
-			let bg_err = self.bg_err.lock();
-			if let Some(err) = &*bg_err {
-				return Err(Error::Background(err.clone()));
-			}
-		}
-
 		let commit: Vec<_> = tx.into_iter().map(
 			|(c, k, v)| (c, self.columns[c as usize].hash(k.as_ref()), v)
 		).collect();
 
+		self.commit_raw(commit)
+	}
+
+	fn commit_raw(&self, commit: Vec<(ColId, Key, Option<Value>)>) -> Result<()> {
 		{
 			let mut queue = self.commit_queue.lock();
 			if queue.bytes > MAX_COMMIT_QUEUE_BYTES {
 				log::debug!(target: "parity-db", "Waiting, qb={}", queue.bytes);
 				self.commit_queue_full_cv.wait(&mut queue);
 			}
+			{
```

Comment:
There was a deadlock possible when the background thread stops because of the error. Commit would wait here indefinitely. Now the error wakes up the commiting thread and the error is checked after that.